### PR TITLE
W3DProjectileStreamDraw and ProjectileStreamUpdate

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/draw/w3dprojectilestreamdraw.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+NameKeyType W3DProjectileStreamDraw::Get_Module_Name_Key() const
+{
+    static NameKeyType nk = g_theNameKeyGenerator->Name_To_Key("W3DProjectileStreamDraw");
+    return nk;
+}
+
+void W3DProjectileStreamDraw::Do_Draw_Module(const Matrix3D *transform)
+{
+    Object *object = Get_Drawable()->Get_Object();
+    static const NameKeyType key_ProjectileStreamUpdate = g_theNameKeyGenerator->Name_To_Key("ProjectileStreamUpdate");
```
Comment (tomsons26) [inline src/game/client/draw/w3dprojectilestreamdraw.cpp]: all this is in a object != nullptr

Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.cpp]: Oops, fixed.

File: src/game/client/draw/w3dprojectilestreamdraw.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    ModuleData::Build_Field_Parse(p);
+    p.Add(dataFieldParse, 0);
+}
+
+W3DProjectileStreamDraw::W3DProjectileStreamDraw(Thing *thing, ModuleData const *module_data) :
+    DrawModule(thing, module_data)
+{
+    m_texture = W3DAssetManager::Get_Instance()->Get_Texture(Get_W3D_Projectile_Stream_Draw_Module_Data()->m_texture);
+
+    for (int i = 0; i < 20; i++) {
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.cpp]: Either use a constant or `ARRAY_SIZE(m_allLines)` for all the 20's in this class. I suggest adding enum value `MAX_SEGMENT_COUNT`.

File: src/game/client/draw/w3dprojectilestreamdraw.cpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+void W3DProjectileStreamDraw::Do_Draw_Module(const Matrix3D *transform)
+{
+    Object *object = Get_Drawable()->Get_Object();
+    if (object != nullptr) {
+        static const NameKeyType key_ProjectileStreamUpdate = g_theNameKeyGenerator->Name_To_Key("ProjectileStreamUpdate");
+        ProjectileStreamUpdate *update = (ProjectileStreamUpdate *)object->Find_Update_Module(key_ProjectileStreamUpdate);
+        W3DProjectileStreamDrawModuleData *data = Get_W3D_Projectile_Stream_Draw_Module_Data();
+
+        Vector3 points[20];
+        int count;
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.cpp]: `projectile_count`

File: src/game/client/draw/w3dprojectilestreamdraw.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        update->Get_All_Points(points, &count);
+
+        Vector3 points2[20];
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.cpp]: `line_points`

File: src/game/client/draw/w3dprojectilestreamdraw.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+void W3DProjectileStreamDraw::Do_Draw_Module(const Matrix3D *transform)
+{
+    Object *object = Get_Drawable()->Get_Object();
+    if (object != nullptr) {
+        static const NameKeyType key_ProjectileStreamUpdate = g_theNameKeyGenerator->Name_To_Key("ProjectileStreamUpdate");
+        ProjectileStreamUpdate *update = (ProjectileStreamUpdate *)object->Find_Update_Module(key_ProjectileStreamUpdate);
+        W3DProjectileStreamDrawModuleData *data = Get_W3D_Projectile_Stream_Draw_Module_Data();
+
+        Vector3 points[20];
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.cpp]: `projectile_points`

File: src/game/client/draw/w3dprojectilestreamdraw.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.cpp]: Used ARRAY_SIZE for this.

File: src/game/client/draw/w3dprojectilestreamdraw.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.cpp]: fixed.

File: src/game/client/draw/w3dprojectilestreamdraw.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.cpp]: fixed.

File: src/game/client/draw/w3dprojectilestreamdraw.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.cpp]: fixed.

File: src/game/client/draw/w3dprojectilestreamdraw.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+void W3DProjectileStreamDraw::Do_Draw_Module(const Matrix3D *transform)
+{
+    Object *object = Get_Drawable()->Get_Object();
+    if (object != nullptr) {
+        static const NameKeyType key_ProjectileStreamUpdate = g_theNameKeyGenerator->Name_To_Key("ProjectileStreamUpdate");
+        ProjectileStreamUpdate *update = (ProjectileStreamUpdate *)object->Find_Update_Module(key_ProjectileStreamUpdate);
+        W3DProjectileStreamDrawModuleData *data = Get_W3D_Projectile_Stream_Draw_Module_Data();
+
+        Vector3 all_points[20];
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.cpp]: `ARRAY_SIZE(m_allLines)` or `MAX_SEGMENT_COUNT`

File: src/game/client/draw/w3dprojectilestreamdraw.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        int points_used;
+        update->Get_All_Points(all_points, &points_used);
+
+        Vector3 staging_points[20];
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.cpp]: `ARRAY_SIZE(m_allLines)` or `MAX_SEGMENT_COUNT`

File: src/game/logic/object/update/projectilestreamupdate.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    static Module *Friend_New_Module_Instance(Thing *thing, ModuleData const *module_data);
+
+    void Add_Projectile(ObjectID source_id, ObjectID projectile_id, ObjectID target_id, const Coord3D *target_pos);
+    void Cull_Front_Of_List();
+    bool Consider_Dying();
+    void Get_All_Points(Vector3 *points, int *count);
+    void Set_Position(Coord3D *new_position);
+
+private:
+    ObjectID m_projectileIDs[20];
```
Comment (xezon) [inline src/game/logic/object/update/projectilestreamupdate.h]: There ideally is a connection between `m_projectileIDs[20]` and `m_allLines[20]`.

File: src/game/client/draw/w3dprojectilestreamdraw.h [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    virtual void CRC_Snapshot(Xfer *xfer) override;
+    virtual void Xfer_Snapshot(Xfer *xfer) override;
+    virtual void Load_Post_Process() override;
+
+    void Make_Or_Update_Line(Vector3 *points, unsigned int point_count, int line_index);
+
+    static ModuleData *Friend_New_Module_Data(INI *ini);
+    static Module *Friend_New_Module_Instance(Thing *thing, ModuleData const *module_data);
+
+    W3DProjectileStreamDrawModuleData *Get_W3D_Projectile_Stream_Draw_Module_Data() const
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.h]: `const W3DProjectileStreamDrawModuleData *`

File: src/game/client/draw/w3dprojectilestreamdraw.h [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        return (W3DProjectileStreamDrawModuleData *)Module::Get_Module_Data();
```
Comment (xezon) [inline src/game/client/draw/w3dprojectilestreamdraw.h]: `static_cast<const W3DProjectileStreamDrawModuleData *>`

File: src/game/logic/object/update/projectilestreamupdate.cpp [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+    m_nextFreeIndex(0),
+    m_firstValidIndex(0),
+    m_targetID(OBJECT_UNK)
+{
+    m_targetPos.Zero();
+}
+
+NameKeyType ProjectileStreamUpdate::Get_Module_Name_Key() const
+{
+    static NameKeyType nk = g_theNameKeyGenerator->Name_To_Key("ProjectileStreamUpdate");
```
Comment (xezon) [inline src/game/logic/object/update/projectilestreamupdate.cpp]: `static const`

File: src/game/client/draw/w3dprojectilestreamdraw.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.cpp]: fixed.

File: src/game/logic/object/update/projectilestreamupdate.h (hunk 8 above)
Comment (jonwil) [reply src/game/logic/object/update/projectilestreamupdate.h]: fixed.

File: src/game/client/draw/w3dprojectilestreamdraw.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.cpp]: fixed.

File: src/game/logic/object/update/projectilestreamupdate.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/object/update/projectilestreamupdate.cpp]: fixed.

File: src/game/client/draw/w3dprojectilestreamdraw.h (hunk 10 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.h]: fixed.

File: src/game/client/draw/w3dprojectilestreamdraw.h (hunk 9 above)
Comment (jonwil) [reply src/game/client/draw/w3dprojectilestreamdraw.h]: fixed.

**Review by xezon — APPROVED** [review]
Code looks good. Projectile Stream works properly in game.

**Pull request merged**